### PR TITLE
feat: protect branch when init github repo

### LINKED ISSUE #963: :four_leaf_clover: `Proposal`: repo scaffolding should protect the default main branch
## What Would You Like to Add? Why Is This Needed?

After repo scaffolding, in the new repo, there is a warning by GitHub: 

Your main branch isn't protected

Protect this branch from force pushing, deletion, or require status checks before merging.

## Design

We should protect the default main branch of a newly created repo in GitHub.

### PR DESCRIPTION
**Pull request by csonezp**
Signed-off-by: csonezp <csonezp@gmail.com>

## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [x] I have added relevant tests

## Description
repo-scaffolding should protect the default main branch.

I'm not used to using ginkgo, so the test cases may not be well designed, If you find any problem please submit some suggestions

## Related Issues
Closes #963 

## New Behavior (screenshots if needed)
When use repo-scaffolding to create destination repo, protected the default branch.

### CODE REVIEW

File: pkg/util/github/repo.go
Comment (iyear) [inline pkg/util/github/repo.go:162]: I looked into github branch protection.

I think here is a configuration we can discuss that
1. `DismissStaleReviews` set to `false`, otherwise the reviewer will frequently approve, resulting in too much trouble to reviewer.
2. set `RequiredApprovingReviewCount` to `1`, so that PRs after approving can be merged, and the user can adjust them later.
3. set `RequiredConversationResolution` to `true`, all comments raised should be resolved.

Full config:
```go
req := &github.ProtectionRequest{
	RequiredPullRequestReviews: &github.PullRequestReviewsEnforcementRequest{
		RequireCodeOwnerReviews:      true,
		RequiredApprovingReviewCount: 1,
	},
	RequiredConversationResolution: github.Bool(true),
}
```

What's your opnion? @csonezp

File: pkg/util/github/repo.go
Comment (iyear) [inline pkg/util/github/repo.go:177]: We probably don't need to judge the Status Code. I didn't find a use for it in the `go-github` examples and resp is usually used for pagination polling.

Comment (iyear) [inline pkg/util/github/repo.go:177]: We don't need to judge the `Status Code`, the `Status Code` has already been judged and converted to `err` in `go-github`, and `resp` is usually used for pagination polling.

https://github.com/google/go-github/blob/master/github/github.go#L997-L1042

**Comment by iyear** [comment]
@csonezp Thank you so much for the contribution, it's an enhancement to github package!

File: pkg/util/github/repo.go
Comment (csonezp) [reply pkg/util/github/repo.go:162]: @iyear  I agree with you except for ```DismissStaleReviews```.
In my option, once a pull request is approved, if there are further commits, the pull request should go to un-approved state automatically.
This avoids unreviewed code being merged into the main branch.
Do I understand right?

File: pkg/util/github/repo.go
Comment (csonezp) [reply pkg/util/github/repo.go:177]: ok, i got it

File: pkg/util/github/repo.go
Comment (iyear) [reply pkg/util/github/repo.go:162]: I agree with you. I've given it some thought, `DismissStaleReviews` is still set to `true`, we just go by the strictest rules and the user adjusts them.

**Comment by csonezp** [comment]
@iyear Issues mentioned in the conversation have been resolved.

File: pkg/util/github/repo.go
Comment (iyear) [inline pkg/util/github/repo.go:166]: If this function returns an error to upstream, do not print the error, as this will cause the error to be printed twice.

All similarities need to be modified. Thank you for your hard work!

Comment (csonezp) [reply pkg/util/github/repo.go]: My bad, Golang is not my most used language, thanks for your suggestion

Comment (iyear) [reply pkg/util/github/repo.go]: line 171 need to be modified as well.

Comment (iyear) [reply pkg/util/github/repo.go]: That's nothing, it's our greatest joy to see community contributions.

**Review by iyear — APPROVED** [review]
Awesome feature!

**Comment by iyear** [comment]
We advocate combining similar content from one PR into one commit, see our [blog post](https://blog.devstream.io/posts/open-a-pr-in-github/#62commits-%E5%A4%AA%E5%A4%9A%E6%88%96%E8%80%85%E8%AE%B0%E5%BD%95%E6%B7%B7%E4%B9%B1%E5%A6%82%E4%BD%95%E5%90%88%E5%B9%B6-commits) for details:

Of course for initial contributors, this is not required. You can take note of this the next time you submit a PR.
.

**Comment by csonezp** [comment]
> We advocate combining similar content from one PR into one commit, see our [blog post](https://blog.devstream.io/posts/open-a-pr-in-github/#62commits-%E5%A4%AA%E5%A4%9A%E6%88%96%E8%80%85%E8%AE%B0%E5%BD%95%E6%B7%B7%E4%B9%B1%E5%A6%82%E4%BD%95%E5%90%88%E5%B9%B6-commits) for details.
> 
> Of course for initial contributors, this is not required. You can take note of this the next time you submit a PR.

In other open source projects I have participated in, the step of squashing commits is done by project members who merge commits, and I didn't find any instructions for squash in the contributor guide, so I keep multiple commits to preserve the review context.

We can add this requirement to the contributor guidelines to prevent other contributors from not knowing this information

**Comment by iyear** [comment]
@csonezp Thank you for the reminder, I will discuss this matter within members.

**Review by daniel-hutao — APPROVED** [review]

**Pull request merged**
